src/preprocessing/point_in_poly/hierarchical_winding.jl

```diff
@@ -0,0 +1,239 @@
+    (; max_corner, min_corner, faces) = bounding_box
+
+    if length(faces) < 100
+        bounding_box.is_leaf[] = true
+
+        return bounding_box
+    end
+
+    determine_closure!(bounding_box, mesh, faces, directed_edges)
+
+    if length(bounding_box.closing_faces) >= length(faces)
+        bounding_box.is_leaf[] = true
+
+        return bounding_box
+    end
+
+    # Bisect the box splitting its longest side
+    box_edges = max_corner - min_corner
+
+    split_direction = findfirst(x -> maximum(box_edges) == x, box_edges)
+
+    uvec = (1:3) .== split_direction
+
+    max_corner_left = max_corner - 0.5box_edges[split_direction] * uvec
+    min_corner_right = min_corner + 0.5box_edges[split_direction] * uvec
+
+    faces_left = in_bounding_box(mesh, faces, min_corner, max_corner_left)
+    faces_right = in_bounding_box(mesh, faces, min_corner_right, max_corner)
+
+    bbox_left = BoundingBoxTree(faces_left, min_corner, max_corner_left)
+    bbox_right = BoundingBoxTree(faces_right, min_corner_right, max_corner)
+
+    push!(bounding_box.children, bbox_left)
+    push!(bounding_box.children, bbox_right)
+
+    construct_hierarchy!(bbox_left, mesh, directed_edges)
+    construct_hierarchy!(bbox_right, mesh, directed_edges)
+
+    return bounding_box
+end
+
+# This only works when all `vertices` are unique
+function determine_closure!(bounding_box, mesh::TriangleMesh{3}, faces, count_directed_edge)
+    (; edge_vertices_ids, face_vertices_ids, face_edges_ids, vertices) = mesh
+    (; min_corner, max_corner) = bounding_box
+
+    count_directed_edge .= 0
+    intersecting_faces = Int[]
+
+    # Find all exterior edges
+    for face in faces
+        v1 = face_vertices_ids[face][1]
+        v2 = face_vertices_ids[face][2]
+        v3 = face_vertices_ids[face][3]
+
+        if in_bounding_box(vertices[v1], min_corner, max_corner) &&
+           in_bounding_box(vertices[v2], min_corner, max_corner) &&
+           in_bounding_box(vertices[v3], min_corner, max_corner)
+            # Face is completely inside the bounding box
+
+            edge1 = face_edges_ids[face][1]
+            edge2 = face_edges_ids[face][2]
+            edge3 = face_edges_ids[face][3]
+
+            if edge_vertices_ids[edge1] == (v1, v2)
+                count_directed_edge[edge1] += 1
+            else
+                count_directed_edge[edge1] -= 1
+            end
+            if edge_vertices_ids[edge2] == (v2, v3)
+                count_directed_edge[edge2] += 1
+            else
+                count_directed_edge[edge2] -= 1
+            end
+            if edge_vertices_ids[edge3] == (v3, v1)
+                count_directed_edge[edge3] += 1
+            else
+                count_directed_edge[edge3] -= 1
+            end
+
+        elseif !in_bounding_box(vertices[v1], min_corner, max_corner) ||
+               !in_bounding_box(vertices[v2], min_corner, max_corner) ||
+               !in_bounding_box(vertices[v3], min_corner, max_corner)
```

Comment:
```suggestion
        else
```
This is literally the negation of the if condition.